spring-integration-core/src/main/java/org/springframework/integration/config/GlobalChannelInterceptorProcessor.java

```diff
@@ -96,6 +100,21 @@ public void afterSingletonsInstantiated() {
 				addMatchingInterceptors(entry.getValue(), entry.getKey());
 			}
 		}
+
+		this.singletonsInstantiated = true;
+	}
+
+	@Override
+	public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+		return bean;
+	}
+
+	@Override
+	public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+		if (this.singletonsInstantiated && bean instanceof ChannelInterceptorAware) {
+			addMatchingInterceptors((ChannelInterceptorAware) bean, beanName);
```

Comment:
I think we need to make this optional somehow, default false - since it's a behavior change - some people might get channels intercepted that they didn't previously - they might be already manually adding their interceptor to channels they register manually.

Reply:
Also, we need to track what channels have been post processed otherwise you'll keep on adding interceptors with each invocation of this operation

Reply:
Not, sure, Oleg, what you mean since this is how `BPP` works.

With the `this.singletonsInstantiated` we are going to apply these global interceptors only to the channels created dynamically, at runtime.

Reply:
Why? This will only be called once per bean.

Reply:
OK, Gary. Then plan is like this:

1. Global Integration property - `false` by default.
2. Remove this property in `5.1` and just rely on the standard `BPP` behavior here.
3. Make this class `public` for Oleg's solution in the https://github.com/spring-cloud/spring-cloud-stream/pull/1227.
4. Leave Oleg's fix in the SCSt since turning on the mentioned property there is going to be a breaking change as well and we are still requested to back-port the fix.

Or another way:

Push this issue back to `5.1` as a normal `BPP` solution without any flags. As we all know there is a workaround - adding these global interceptors manually. In this case we may leave for the fix in the SCSt for a while.

(I might missed something...)

Reply:
@artembilan in your test call this twice 
```
beanFactory.initializeBean(testChannel, "testChannel");
```
and you'll see what I mean.

Reply:
>Make this class public

Why? SCSt can simply set the property to true and it will all work.

>call this twice

That's not really a valid thing to do; I don't think we have to protect against foolishness.

Reply:
@artembilan Oh; I see what you mean - if SCSt sets the flag to true their own channels might be double intercepted - OK.

Reply:
Correct. That's why I say that the flag to `true` in SCSt is a breaking change as well.
So, if we are not going this "let it break anyway!" way, the only solution as I've chosen in the beginning - move the fix to `5.1` and leave with the workaround in SCSt which is clearly applied only to the internally created channels by the `BinderAwareRouterBeanPostProcessor`.

Reply:
No; I am concurring with your 1,2,3,4 plan above.

Reply:
Good. I'll also add a note to the `Global Channel Interceptor Configuration` section to describe how it works now and what option should be turned on otherwise.